[OpInfo] Add OpInfo tests for torch.inner

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55536 [OpInfo] Add OpInfo tests for torch.inner**

Differential Revision: [D27650199](https://our.internmc.facebook.com/intern/diff/D27650199)

ejguan committed:
Add OpInfo tests for torch.inner

[ghstack-poisoned]

ejguan committed:
Update on "Add OpInfo tests for torch.inner"




[ghstack-poisoned]

ejguan committed:
Update on "Add OpInfo tests for torch.inner"




[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3876,6 +3892,16 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           dtypesIfROCM=floating_types_and(torch.half),
+           supports_out=True,
+           sample_inputs_func=sample_inputs_inner,
+           skips=(SkipInfo('TestCommon', 'test_out', dtypes=[torch.float32]),)),
```
Would you add a comment explaining why test_out has to be skipped?

ejguan replied:
A comment is attached with the issue.

ejguan committed:
Update on "Add OpInfo tests for torch.inner"


Differential Revision: [D27650199](https://our.internmc.facebook.com/intern/diff/D27650199)

[ghstack-poisoned]

ejguan committed:
Update on "Add OpInfo tests for torch.inner"


Differential Revision: [D27650199](https://our.internmc.facebook.com/intern/diff/D27650199)

[ghstack-poisoned]

ejguan commented:
The CI has a bunch of errors not related to this PR.

mruberry reviewed (approved):
Awesome! Thanks @ejguan!

ejguan committed:
Update on "Add OpInfo tests for torch.inner"


Differential Revision: [D27650199](https://our.internmc.facebook.com/intern/diff/D27650199)

[ghstack-poisoned]

ejguan committed:
Update on "[OpInfo] Add OpInfo tests for torch.inner"


Differential Revision: [D27650199](https://our.internmc.facebook.com/intern/diff/D27650199)

[ghstack-poisoned]

imaginary-person commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4402,6 +4418,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           # BFloat16 support on CUDA requires CUDA 11 and SM53
+           dtypesIfCUDA=floating_types_and(torch.float16, torch.complex64, torch.complex128,
+                                           *[torch.bfloat16] if CUDA11OrLater else []),
+           dtypesIfROCM=floating_types_and(torch.half),
```
I don't see `inner` in `derivatives.yaml`. Does that mean it doesn't support autograd?

mruberry replied:
Hey @imaginary-person! That's a good question, and we have to look at native_functions.yaml to know the answer:

https://github.com/pytorch/pytorch/blob/85126629a59ac51dcebca08b0ef2da5dd3afd827/aten/src/ATen/native/native_functions.yaml#L8730

Here we see that inner has no `DISPATCH` entry in its yaml, which means it acquires the default `DISPATCH` type of CompositeImplicitAutograd (see the README [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md)). 

Basically, we can think of "CompositeImplicitAutograd" as letting inner's autograd formula be defined by the operations its composed of, so it doesn't need a native_functions.yaml entry to support autograd.

"CompositeExplicitAutograd," by the way, means the operation is still composite but implements its own autograd. So all functions that support autograd either have derivatives.yaml entries or are CompositeImplicitAutograd dispatch types.

Does that make sense? 

Now as for testing, it's true that there may be some cases for complicit operations where testing there behavior is redundant with testing the behavior of the ops they're composed of. However, determining this would take some more work, and it would be tricky to keep up to date with operator implementations. If the implementation of inner changed, for example, maybe we would need to carefully tweak its test surface again. So for now we try to be (mostly) agnostic about how operators are implemented OR, if we do have implementation-specific tests, we try to make them fail if an operator's implementation changes.

imaginary-person replied:
Thanks a lot for the clarification, @mruberry!
